### PR TITLE
[EditableText] call `onSelectionChanged` only when there are actual selection/cause changes

### LINKED ISSUE #76349: EditableText calls onSelectionChanged even if the selection does not change.
If you add a print statement in this method https://github.com/flutter/flutter/blob/934dfde42f3f2c8c7665f9874812f7c35d846ff4/packages/flutter/lib/src/material/text_field.dart#L1070

long press the textfield, you will see the print statement even if text selection does not change. This behavior is contradicted to the EditableText.onSelectionChange contract. Unfortunately, this is needed because textfield relies on this call to update the handle visibility.

We should fix this by refactoring out selection overlay of editable text so that text field does not need to rely on this callback, and then we should finally able to fix the EditableText.onSelectionChanged.

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Fixes https://github.com/flutter/flutter/issues/76349.

Also cleaned up the `updateEditingValue`/`userUpdateEditingValue`
paths a bit. Hopefully we will be able to merge the two.

*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -593,7 +593,6 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
   bool _wasSelectingVerticallyWithKeyboard = false;
 
   void _setTextEditingValue(TextEditingValue newValue, SelectionChangedCause cause) {
-    textSelectionDelegate.textEditingValue = newValue;
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/editable.dart:596]: I kept this for backward compatibility, we should only remove it once we remove the deprecated setter textEditingValue

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 2]
```diff
@@ -1533,9 +1536,15 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
+  // The source of the most recent selection change.
+  //
+  // Changing the selection programmatically does not update
+  // _selectionChangedCause.
+  SelectionChangedCause? _selectionChangedCause;
+
   ScrollController? _scrollController;
 
-  late AnimationController _cursorBlinkOpacityController;
+  late final AnimationController _cursorBlinkOpacityController = AnimationController(vsync: this, duration: _fadeDuration);
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/editable_text.dart:1547]: I think we missed this but it should be disposed in the dispose method.
same for _floatingCursorResetController. I think the existing format is cleaner, it is weird the we give a default value in here and then attach listener in initState

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 3]
```diff
@@ -1748,17 +1756,22 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
-    if (value == _value) {
-      // This is possible, for example, when the numeric keyboard is input,
-      // the engine will notify twice for the same value.
-      // Track at https://github.com/flutter/flutter/issues/65811
-      return;
+    final bool shouldShowCaret = widget.readOnly
+      ? _value.selection != value.selection
+      : _value != value;
+    if (shouldShowCaret) {
+      _scheduleShowCaretOnScreen();
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/editable_text.dart:1763]: I didnt get the logic, the _scheduleShowCaretOnScreen will be called in line 1770 anyway?

File: packages/flutter/test/cupertino/text_field_test.dart [hunk 4]
```diff
@@ -2381,19 +2381,22 @@ void main() {
     await gesture.moveBy(const Offset(600, 0));
     // To the edge of the screen basically.
     await tester.pump();
+    await tester.pumpAndSettle();
```
Comment (chunhtai) [inline packages/flutter/test/cupertino/text_field_test.dart:2384]: where do we introduce frame delays?

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 5]
```diff
@@ -2533,7 +2543,24 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
+    if (value.selection != _value.selection) {
+      requestKeyboard();
+    }
+
+    if (value.text != _value.text || value.composing != _value.composing) {
+      hideToolbar();
+      _currentPromptRectRange = null;
+
+      if (_hasInputConnection) {
+        if (widget.obscureText && value.text.length == _value.text.length + 1) {
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:2557]: I got really worried about extended grapheme clusters and obscured fields when I saw this `+ 1` here (even though I know it has nothing to do with this PR).  Indeed they show up as multiple characters when entered into an obscured field and don't have a delay, but it turns out that's the behavior on native too 🤷 .  All is well.

**Review by justinmc — APPROVED** [review]
LGTM 👍 .  This was badly needed, thanks for cleaning this up.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:1547]: These should be fixed in https://github.com/flutter/flutter/pull/87973

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 3 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:1763]: Ah I forgot to remove the existing `_sheduleShowCaretOnScreen` when copied the other one over.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 6]
```diff
@@ -4940,21 +4940,19 @@ void main() {
 
   testWidgets('keyboard text selection works (RawKeyEvent)', (WidgetTester tester) async {
     debugKeyEventSimulatorTransitModeOverride = KeyDataTransitMode.rawKeyData;
+    addTearDown(() { debugKeyEventSimulatorTransitModeOverride = null; });
 
     await testTextEditing(tester, targetPlatform: defaultTargetPlatform);
-
     debugKeyEventSimulatorTransitModeOverride = null;
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/test/widgets/editable_text_test.dart:4946]: This doesn't seem to work, `_verifyInvariants` is called before the the tear down callbacks are invoked.

File: packages/flutter/test/cupertino/text_field_test.dart (hunk 4 above)
Comment (LongCatIsLooong) [reply packages/flutter/test/cupertino/text_field_test.dart:2384]: There's no need for these changes. Removed.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:596]: Could you elaborate? People are required to implement `userUpdateTextEditingValue` right? It's part of the TextSelectionDelegate interface.

Comment (chunhtai) [reply packages/flutter/lib/src/rendering/editable.dart:596]: I didn't notice userUpdateTextEditingValue is abstract method. yeah this should be fine.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 7]
```diff
@@ -2481,7 +2489,6 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       Rect? composingRect = renderEditable.getRectForComposingRange(composingRange);
       // Send the caret location instead if there's no marked text yet.
       if (composingRect == null) {
-        assert(!composingRange.isValid || composingRange.isCollapsed);
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:2484]: I must have thought `getRectForComposingRange` only returns null only when `!composingRange.isValid || composingRange.isCollapsed` is true but if `getBoxesForSelection` returns an empty list that method returns null too. It's triggered in one of the tests so I removed it.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 8]
```diff
@@ -1880,9 +1881,13 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     final Offset finalPosition = renderEditable.getLocalRectForCaret(_lastTextPosition!).centerLeft - _floatingCursorOffset;
     if (_floatingCursorResetController.isCompleted) {
       renderEditable.setFloatingCursor(FloatingCursorDragState.End, finalPosition, _lastTextPosition!);
-      if (_lastTextPosition!.offset != renderEditable.selection!.baseOffset)
+      if (_lastTextPosition!.offset != renderEditable.selection!.baseOffset) {
         // The cause is technically the force cursor, but the cause is listed as tap as the desired functionality is the same.
-        _handleSelectionChanged(TextSelection.collapsed(offset: _lastTextPosition!.offset), SelectionChangedCause.forcePress);
+        final TextEditingValue newValue = _value.copyWith(
+          selection: TextSelection.fromPosition(_lastTextPosition!),
+        );
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:1883]: The floating cursor stuff currently doesn't care much about affinity (e.g., `final TextPosition currentTextPosition = TextPosition(offset: renderEditable.selection!.baseOffset);`). I'm hoping we can take cursor handling out of `EditableTextState` and maybe clean floating cursor a bit in the process.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart]: moved the `addListener` to the lazy initializer. My initial goal was to mark the property as final. I'll do the same for other animation controllers in #87973.

File: packages/flutter/test/widgets/editable_text_test.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/test/widgets/editable_text_test.dart:4946]: Ah ok interesting, best to keep both then 👍

**Pull request merged**